bin/sam-coverage-depth.py

```diff
@@ -2,169 +2,391 @@
+    baseCounts = []
+    writerow = csv.writer(sys.stdout, delimiter=args.sep).writerow if args.tsv else None
+
+    if writerow:
+        row = ["Site"]
+        if referenceSeq:
+            row.extend(("Reference", "Mutations"))
+        writerow(row + list(BASES))
+
+    readIds = set()
+    totalBases = totalMutationCount = 0
+    # I don't use a collections.Counter in the following because they are very slow.
+    mutationPairCounts = defaultdict(int)
+    start = time()
+    totalRunTime = maxRunTime = 0.0
+    startSites = list(range(0, referenceLen, args.window))
+
+    with ProcessPoolExecutor(max_workers=args.workers) as executor:
+        for result in executor.map(
+            processSubsection,
+            repeat(args),
+            startSites,
+            repeat(referenceId),
+            repeat(referenceSeq),
+            repeat(bool(writerow)),
+            repeat(start),
+        ):
+            waited = result["startTime"] - start
+            runTime = result["stopTime"] - result["startTime"]
+            totalRunTime += runTime
+            if runTime > maxRunTime:
+                maxRunTime = runTime
+
+            if args.verbosity > 0:
+                from_ = result['startSite']
+                to = min(from_ + args.window, referenceLen)
+                print(
+                    f"Genome region {from_ + 1}-{to} processed in {runTime:.2f} "
+                    f"seconds, after waiting {waited:.2f} seconds to start.",
+                    file=sys.stderr,
+                )
+
+            func = writerow or print
+            for row in result["rows"]:
+                func(row)
+
+            readIds.update(result["readIds"])
+            totalBases += result["totalBases"]
+            baseCounts.extend(result["baseCounts"])
+            for pair, count in result["mutationPairCounts"].items():
+                mutationPairCounts[pair] += count
+            totalMutationCount += result["totalMutationCount"]
+
+    stop = time()
+    try:
+        speedup = totalRunTime / (stop - start)
+    except ZeroDivisionError:
+        speedup = 0.0
+
+    if args.verbosity > 0:
         print(
-            "Min coverage depth: %d"
-            % (0 if len(counts) < referenceLength else min(counts))
+            f"Finised in {int(stop - start)} seconds. Total serial processing would "
+            f"have been {int(totalRunTime)} seconds. The slowest process took "
+            f"{maxRunTime:.2f} seconds. Speed up = {speedup:.2f}",
+            file=sys.stderr,
         )
-        print("Max coverage depth: %d" % max(counts, default=0))
-        print("Mean coverage depth: %.3f" % (sum(counts) / referenceLength))
-        if counts:
-            print("Coverage depth s.d.: %.3f" % std(counts))
+
+    if args.printStats:
+        out = open(args.statsFile, "w") if args.statsFile else sys.stdout
+        with redirect_stdout(out):
+            referenceLength = referenceLens[referenceId]
```

Comment:
Why are you not using the `referenceLen` you defined on line 292?

Reply:
Because I have no short-term memory?  Fixed.